cpanel-dynamic-dns

```diff
@@ -177,41 +182,23 @@ generate_auth_string () {
 fetch_zone () {
     say "Fetching zone for %s...." "$DOMAIN"
     LAST_CONNECT_HOST=$CPANEL_SERVER
-    REQUEST="GET /xml-api/cpanel?cpanel_xmlapi_module=ZoneEdit&cpanel_xmlapi_func=fetchzone&cpanel_xmlapi_apiversion=2&domain=$DOMAIN HTTP/1.0\r\nConnection: close\r\nAuthorization: Basic $AUTH_STRING\r\nUser-Agent: cpanel-dynamic-dns.sh $VERSION\r\n\r\n\r\n"
+    REQUEST="GET /json-api/cpanel?cpanel_jsonapi_module=ZoneEdit&cpanel_jsonapi_func=fetchzone&cpanel_jsonapi_apiversion=2&domain=$DOMAIN HTTP/1.0\r\nConnection: close\r\nAuthorization: Basic $AUTH_STRING\r\nUser-Agent: cpanel-dynamic-dns $VERSION\r\n\r\n\r\n"
     RECORD=""
     LINES=""
     INRECORD=0
     USETHISRECORD=0
     REQUEST_RESULTS=`printf "$REQUEST" | openssl s_client -quiet -connect $CPANEL_SERVER:2083 2>/dev/null`
 
     check_results_for_error "$REQUEST_RESULTS" "$REQUEST"
-    for LINE in $REQUEST_RESULTS
-    do
-        if [ "$LINE" = "<record>" ]; then
-            INRECORD=1
-            continue
-        fi
-        if [ "$LINE" = "</record>" ]; then
-            INRECORD=0
-            if [ "$USETHISRECORD" = "2" ]; then
-                LINENUM=`printf "$RECORD" | grep '<Line>' | awk -F'<' '{print \$2}' | awk -F'>' '{print \$2}'`
-                ADDRESS=`printf "$RECORD" | grep -i '<address>' | awk -F'<' '{print \$2}' | awk -F'>' '{print \$2}'`
-                LINES="$LINES\n$LINENUM=$ADDRESS"
-            fi
-            USETHISRECORD=0
-            RECORD=""
-            continue
-        fi
-        if [ "$LINE" = "<type>A</type>" ]; then
-            USETHISRECORD=`expr $USETHISRECORD + 1`
-        fi
-        if [ "$LINE" = "<name>$SUBDOMAIN$DOMAIN.</name>" ]; then
-            USETHISRECORD=`expr $USETHISRECORD + 1`
-        fi
-        if [ "$INRECORD" = "1" ]; then
-            RECORD="$RECORD\n$LINE"
-        fi
-    done
+    LINES="$(extract_from_json "$REQUEST_RESULTS" "
+        do {
+            join(qq{\n}, map {
+                qq{\$_->{Line}=\$_->{address}}
+            } grep {
+                \$_->{type} eq q{A} && \$_->{name} eq q{$SUBDOMAIN$DOMAIN.}
+            } @{\$_->{cpanelresult}{data}[0]{record}});
+        }
+    ")"
```

Comment:
Minor: Embedding Perl in a command line argument requires careful escaping.  It might be worthwhile to look into a JSON processor, like `jq` (though it is *not* installed on systems by default).  E.g.

```sh
echo -n "$REQUEST_RESULTS" \
| sed '1,/^\r$/d' \
| jq -r '
    .cpanelresult.data[0].record[]
    | select(.type == "A" and .name == "'$SUBDOMAIN$DOMAIN'.")
    | .Line + "=" + .address
'
```

Reply:
Actually, I suspect `JSON::PP` is commonly installed on systems.  In fact, it's the only JSON module that comes installed on my macOS.

It's probably best to keep this as-is.